Terraform v0.13.0 Panic When Using Module Composition

### Terraform Version

```
> terraform_git -v
Terraform v0.13.0-dev
commit 6fbd3942ea6a2e456d43467a64d34fe81dd08274
```

### Terraform Configuration Files

I have had to just add crash.log to this issue report. This is an interesting panic, as it seems to be that if I have greater than two modules called from a root module I cause this panic. I can always try and strip out internal configuration to produce a test case so I am hoping crash.log will help you debug this panic.

If required I could tar up the config but it is multiple modules all in multiple git repos so hoping the stack trace will suffice.

### Crash Output

https://gist.github.com/adamhorden/9730f0415d5166844ee1a947d8486f81

**panic: no expansion has been registered for module.he_aws_tf_mod_he_aws_tf_app_itf.module.he_aws_tf_mod_he_aws_tf_app_itf_stk_be.module.he_aws_tf_mod_aws_cc_app_itf_stk**

Thanks for reporting this, @adamhorden!

Here's the isolated stack trace, for easier reference by someone working on this bug later:

```
panic: no expansion has been registered for module.he_aws_tf_mod_he_aws_tf_app_itf.module.he_aws_tf_mod_he_aws_tf_app_itf_stk_be.module.he_aws_tf_mod_aws_cc_app_itf_stk

goroutine 648 [running]:
github.com/hashicorp/terraform/instances.(*expanderModule).moduleInstances(0xc0011804e0, 0xc0000950d0, 0x1, 0x1, 0xc0018a8b50, 0x2, 0x4, 0xc0013981a0, 0x29349c0, 0xc000d1f140)
	/Users/adhorden/adhorden/terraform_git/instances/expander.go:267 +0xb72
github.com/hashicorp/terraform/instances.(*expanderModule).moduleInstances(0xc001180320, 0xc0000950c0, 0x2, 0x2, 0xc0015a3b50, 0x1, 0x4, 0xc00182a0a0, 0x9d, 0x2200000029)
	/Users/adhorden/adhorden/terraform_git/instances/expander.go:279 +0x2c9
github.com/hashicorp/terraform/instances.(*expanderModule).moduleInstances(0xc000d1e320, 0xc0000950b0, 0x3, 0x3, 0xc0015a3b50, 0x0, 0x4, 0x44db616, 0x40, 0x36)
	/Users/adhorden/adhorden/terraform_git/instances/expander.go:279 +0x2c9
github.com/hashicorp/terraform/instances.(*Expander).ExpandModule(0xc000d1e340, 0xc0000950b0, 0x3, 0x3, 0x0, 0x0, 0x0)
	/Users/adhorden/adhorden/terraform_git/instances/expander.go:116 +0x108
github.com/hashicorp/terraform/terraform.(*nodeExpandModuleVariable).DynamicExpand(0xc000db8400, 0x2fae6e0, 0xc00108c750, 0x1, 0x1, 0x1)
	/Users/adhorden/adhorden/terraform_git/terraform/node_module_variable.go:42 +0x94
github.com/hashicorp/terraform/terraform.(*Graph).walk.func1(0x29169c0, 0xc000db8400, 0x0, 0x0, 0x0)
	/Users/adhorden/adhorden/terraform_git/terraform/graph.go:84 +0x4d8
github.com/hashicorp/terraform/dag.(*Walker).walkVertex(0xc000e330e0, 0x29169c0, 0xc000db8400, 0xc000e0fb00)
	/Users/adhorden/adhorden/terraform_git/dag/walk.go:387 +0x357
created by github.com/hashicorp/terraform/dag.(*Walker).Update
	/Users/adhorden/adhorden/terraform_git/dag/walk.go:309 +0x11d7
```

I understand that this is a complicated configuration that's hard for you to reduce into a minimal reproducible example. As a compromise, if you could share some details about the heirarchy of modules that you have that might help to understand what we can see in the logs and the stack trace.

One way to quickly get an easily-shareable summary of the modules would be to run `terraform init -upgrade` and then share its full output, which includes summary information about the modules and providers that this configuration is using.

Looks like I am seeing the same:

```
Terraform v0.13.0-beta1
$ tf init
Initializing modules...

Initializing the backend...

Initializing provider plugins...
- Finding hashicorp/aws versions matching ">= 2.52.0, ~> 2.53, ~> 2.53"...
- Finding hashicorp/kubernetes versions matching ">= 1.11.1"...
- Finding hashicorp/local versions matching ">= 1.4.*"...
- Finding hashicorp/null versions matching ">= 2.1.*"...
- Finding hashicorp/random versions matching ">= 2.1.*"...
- Installing hashicorp/aws v2.64.0...
- Installed hashicorp/aws v2.64.0 (signed by HashiCorp)
- Installing hashicorp/kubernetes v1.11.3...
- Installed hashicorp/kubernetes v1.11.3 (signed by HashiCorp)
- Installing hashicorp/local v1.4.0...
- Installed hashicorp/local v1.4.0 (signed by HashiCorp)
- Installing hashicorp/null v2.1.2...
- Installed hashicorp/null v2.1.2 (signed by HashiCorp)
- Installing hashicorp/random v2.2.1...
- Installed hashicorp/random v2.2.1 (signed by HashiCorp)
```

```
panic: no expansion has been registered for module.cluster

goroutine 1765 [running]:
github.com/hashicorp/terraform/instances.(*expanderModule).moduleInstances(0xc0021d0080, 0xc000a80720, 0x2, 0x2, 0xc00235eb50, 0x0, 0x4, 0x10ee3b8, 0x0, 0x32)
	/home/circleci/project/project/instances/expander.go:267 +0xb72
github.com/hashicorp/terraform/instances.(*Expander).ExpandModule(0xc0021d00a0, 0xc000a80720, 0x2, 0x2, 0x0, 0x0, 0x0)
	/home/circleci/project/project/instances/expander.go:116 +0x108
github.com/hashicorp/terraform/terraform.(*nodeExpandModuleVariable).DynamicExpand(0xc000e43dc0, 0x2face20, 0xc001c9c8f0, 0x1, 0x1, 0x1)
	/home/circleci/project/project/terraform/node_module_variable.go:42 +0x94
github.com/hashicorp/terraform/terraform.(*Graph).walk.func1(0x2916760, 0xc000e43dc0, 0x0, 0x0, 0x0)
	/home/circleci/project/project/terraform/graph.go:84 +0x4d8
github.com/hashicorp/terraform/dag.(*Walker).walkVertex(0xc0006541e0, 0x2916760, 0xc000e43dc0, 0xc0029c9bc0)
	/home/circleci/project/project/dag/walk.go:387 +0x357
created by github.com/hashicorp/terraform/dag.(*Walker).Update
	/home/circleci/project/project/dag/walk.go:309 +0x11d7
```

Team,

We are experiencing the same/similar issue with one of our modules (using terraform version v0.13.0-beta1)

Here is the full crash.log https://gist.github.com/ficofer/3350a93db889f2aa64b9c8b41a2590c4

This is the short output:

```
2020/06/04 19:18:09 [DEBUG] Starting graph walk: walkValidate
2020/06/04 19:18:09 [TRACE] dag/walk: visiting "module.prometheus.module.thanos_query_cluster.local.tcp_protocol (expand)"
2020/06/04 19:18:09 [TRACE] vertex "module.prometheus.module.thanos_query_cluster.local.tcp_protocol (expand)": starting visit (*terraform.nodeExpandLocal)
2020/06/04 19:18:09 [TRACE] vertex "module.prometheus.module.thanos_query_cluster.local.tcp_protocol (expand)": expanding dynamic subgraph
2020/06/04 19:18:09 [TRACE] vertex "module.prometheus.module.thanos_query_cluster.local.tcp_protocol (expand)": visit complete
panic: no expansion has been registered for module.prometheus

goroutine 263 [running]:
github.com/hashicorp/terraform/instances.(*expanderModule).moduleInstances(0xc0005da320, 0xc000360280, 0x2, 0x2, 0xc000855b20, 0x0, 0x4, 0x7eb68b0, 0x10, 0x4)
	/home/circleci/project/project/instances/expander.go:267 +0xb72
github.com/hashicorp/terraform/instances.(*Expander).ExpandModule(0xc0005da340, 0xc000360280, 0x2, 0x2, 0x0, 0x0, 0x0)
	/home/circleci/project/project/instances/expander.go:116 +0x108
github.com/hashicorp/terraform/terraform.(*nodeExpandLocal).DynamicExpand(0xc000ac6300, 0x2face20, 0xc000a64000, 0x1, 0x1, 0x1)
	/home/circleci/project/project/terraform/node_local.go:70 +0x94
github.com/hashicorp/terraform/terraform.(*Graph).walk.func1(0x28d24a0, 0xc000ac6300, 0x0, 0x0, 0x0)
	/home/circleci/project/project/terraform/graph.go:84 +0x4d8
github.com/hashicorp/terraform/dag.(*Walker).walkVertex(0xc000178c60, 0x28d24a0, 0xc000ac6300, 0xc00065c000)
	/home/circleci/project/project/dag/walk.go:387 +0x357
created by github.com/hashicorp/terraform/dag.(*Walker).Update
	/home/circleci/project/project/dag/walk.go:309 +0x11d7
```

The module causing this error has a child module that calls a third module with _for_each_

Hi. I believe that the crash I am experiencing in my code is related to this bug report, because the panic message looks similar.

This Bash command sequence:

````
$ git clone -b terraform-v0.13beta1-test https://github.com/amg1127/MyCloudInfrastructure
$ cd MyCloudInfrastructure
$ terraform init
$ terraform plan
````

triggers the following Terraform crash:

````
panic: no expansion has been registered for module.prodInfrastructure

goroutine 560 [running]:
github.com/hashicorp/terraform/instances.(*expanderModule).moduleInstances(0xc00046b600, 0xc00009c400, 0x2, 0x2, 0xc0008bfb20, 0x0, 0x4, 0xc000bf0000, 0x93, 0x50000)
	/home/circleci/project/project/instances/expander.go:267 +0xb72
github.com/hashicorp/terraform/instances.(*Expander).ExpandModule(0xc00046b620, 0xc00009c400, 0x2, 0x2, 0x0, 0x0, 0x0)
	/home/circleci/project/project/instances/expander.go:116 +0x108
github.com/hashicorp/terraform/terraform.(*nodeExpandLocal).DynamicExpand(0xc00047d3e0, 0x23ae440, 0xc00062d110, 0x1, 0x1, 0x47d001)
	/home/circleci/project/project/terraform/node_local.go:70 +0x94
github.com/hashicorp/terraform/terraform.(*Graph).walk.func1(0x1cd4400, 0xc00047d3e0, 0x0, 0x0, 0x0)
	/home/circleci/project/project/terraform/graph.go:84 +0x4d8
github.com/hashicorp/terraform/dag.(*Walker).walkVertex(0xc00050c720, 0x1cd4400, 0xc00047d3e0, 0xc000797540)
	/home/circleci/project/project/dag/walk.go:387 +0x357
created by github.com/hashicorp/terraform/dag.(*Walker).Update
	/home/circleci/project/project/dag/walk.go:309 +0x11d7



!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!

Terraform crashed! This is always indicative of a bug within Terraform.
A crash log has been placed at "crash.log" relative to your current
working directory. It would be immensely helpful if you could please
report the crash with Terraform[1] so that we can fix this.

When reporting bugs, please include your terraform version. That
information is available on the first line of crash.log. You can also
get it by running 'terraform --version' on the command line.

SECURITY WARNING: the "crash.log" file that was created may contain 
sensitive information that must be redacted before it is safe to share 
on the issue tracker.

[1]: https://github.com/hashicorp/terraform/issues

!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!
````

The complete `crash.log` file is published here: https://gist.github.com/amg1127/f17c8ae7f6fb3dee6820df5edd6402ff

The Terraform version is:

````
$ ../terraform -v
Terraform v0.13.0-beta1
+ provider registry.terraform.io/hashicorp/null v2.1.2
+ provider registry.terraform.io/hashicorp/random v2.2.1
+ provider registry.terraform.io/terraform-providers/docker v2.7.1
````

--- a/terraform/context_validate_test.go
+++ b/terraform/context_validate_test.go
@@ -1552,3 +1552,83 @@ resource "aws_instance" "foo" {
 		t.Fatalf("wrong error:\ngot:  %s\nwant: message containing %q", got, want)
 	}
 }
+
+func TestContext2Validate_expandMultipleNestedModules(t *testing.T) {
+	m := testModuleInline(t, map[string]string{
+		"main.tf": `
+module "modA" {
+  for_each = {
+    first = "m"
+	second = "n"
+  }
+  source = "./modA"
+}
+`,
+		"modA/main.tf": `
+locals {
+  m = {
+    first = "m"
+	second = "n"
+  }
+}
+
+module "modB" {
+  for_each = local.m
+  source = "./modB"
+  y = each.value
+}
+
+module "modC" {
+  for_each = local.m
+  source = "./modC"
+  x = module.modB[each.key].out
+  y = module.modB[each.key].out
+}
+
+`,
+		"modA/modB/main.tf": `
+variable "y" {
+  type = string
+}
+
+resource "aws_instance" "foo" {
+  foo = var.y
+}
+
+output "out" {
+  value = aws_instance.foo.id
+}
+`,
+		"modA/modC/main.tf": `
+variable "x" {
+  type = string
+}
+
+variable "y" {
+  type = string
+}
+
+resource "aws_instance" "foo" {
+  foo = var.x
+}
+
+output "out" {
+  value = var.y
+}
+`,
+	})
+
+	p := testProvider("aws")
+	p.DiffFn = testDiffFn
+	ctx := testContext2(t, &ContextOpts{
+		Config: m,
+		Providers: map[addrs.Provider]providers.Factory{
+			addrs.NewDefaultProvider("aws"): testProviderFuncFixed(p),
+		},
+	})
+
+	diags := ctx.Validate()
+	if diags.HasErrors() {
+		t.Fatal(diags.ErrWithWarnings())
+	}
+}

--- a/terraform/node_module_variable.go
+++ b/terraform/node_module_variable.go
@@ -150,7 +150,7 @@ func (n *nodeModuleVariable) Path() addrs.ModuleInstance {
 
 // GraphNodeModulePath
 func (n *nodeModuleVariable) ModulePath() addrs.Module {
-	return n.Addr.Module.Parent().Module()
+	return n.Addr.Module.Module()
 }
 
 // RemovableIfNotTargeted

--- a/terraform/node_output.go
+++ b/terraform/node_output.go
@@ -22,6 +22,7 @@ var (
 	_ RemovableIfNotTargeted     = (*nodeExpandOutput)(nil)
 	_ GraphNodeReferenceable     = (*nodeExpandOutput)(nil)
 	_ GraphNodeReferencer        = (*nodeExpandOutput)(nil)
+	_ GraphNodeReferenceOutside  = (*nodeExpandOutput)(nil)
 	_ GraphNodeDynamicExpandable = (*nodeExpandOutput)(nil)
 	_ graphNodeTemporaryValue    = (*nodeExpandOutput)(nil)
 	_ graphNodeExpandsInstances  = (*nodeExpandOutput)(nil)

--- a/terraform/transform_module_expansion.go
+++ b/terraform/transform_module_expansion.go
@@ -120,8 +120,6 @@ func (t *ModuleExpansionTransformer) transform(g *Graph, c *configs.Config, pare
 
 		case GraphNodeModulePath:
 			path = t.ModulePath()
-		case GraphNodeReferenceOutside:
-			path, _ = t.ReferenceOutside()
 		default:
 			continue
 		}
@@ -134,7 +132,9 @@ func (t *ModuleExpansionTransformer) transform(g *Graph, c *configs.Config, pare
 
 	// Also visit child modules, recursively.
 	for _, cc := range c.Children {
-		return t.transform(g, cc, v)
+		if err := t.transform(g, cc, v); err != nil {
+			return err
+		}
 	}
 
 	return nil

